Guide: Connect a Micronaut JMS Application to an AWS SQS Queue

sbodvanski opened the pull request.

sbodvanski committed:
Guide: Connect a Micronaut JMS Application to an AWS SQS Queue

burtbeckwith commented on guides/micronaut-jms-aws-sqs/groovy/src/main/groovy/example/micronaut/Application.groovy:
```diff
@@ -0,0 +1,11 @@
+package example.micronaut
```
delete all 3 `Application` classes - these are generated with the rest of the initial app. You only need to add new or modified source files

sbodvanski replied:
Thank you. That's good to know.

burtbeckwith commented on guides/micronaut-jms-aws-sqs/java/src/main/java/example/micronaut/DemoProducer.java:
```diff
@@ -0,0 +1,14 @@
+public interface DemoProducer {
+
+    @Queue("demo_queue")
+    public void send(@MessageBody String body);
```
remove `public`

burtbeckwith commented on guides/micronaut-jms-aws-sqs/groovy/src/main/groovy/example/micronaut/DemoConsumer.groovy:
```diff
@@ -0,0 +1,19 @@
+
+@JMSListener(CONNECTION_FACTORY_BEAN_NAME)
+class DemoConsumer {
+    private static Logger LOGGER = LoggerFactory.getLogger(DemoConsumer.class);
```
remove this and the 2 imports and annotate the class with `@Slf4j('LOGGER')`

burtbeckwith commented on guides/micronaut-jms-aws-sqs/kotlin/src/main/kotlin/example/micronaut/SqsClientFactory.kt:
```diff
@@ -0,0 +1,22 @@
+    fun sqsClient(environment: Environment?): AmazonSQS {
+        return AmazonSQSClientBuilder
+            .standard()
+            .withRegion(Regions.US_EAST_1)
```
use a static import to be consistent with the Java and Groovy classes

burtbeckwith commented on guides/micronaut-jms-aws-sqs/kotlin/src/main/kotlin/example/micronaut/SqsClientFactory.kt:
```diff
@@ -0,0 +1,22 @@
+class SqsClientFactory {
+
+    @Singleton
+    fun sqsClient(environment: Environment?): AmazonSQS {
```
remove `?`

burtbeckwith commented on guides/micronaut-jms-aws-sqs/kotlin/src/main/kotlin/example/micronaut/DemoProducer.kt:
```diff
@@ -0,0 +1,13 @@
+interface DemoProducer {
+
+    @Queue("demo_queue")
+    fun send(@MessageBody body: String?)
```
remove `?` unless null body is allowed

burtbeckwith commented on guides/micronaut-jms-aws-sqs/kotlin/src/main/kotlin/example/micronaut/DemoConsumer.kt:
```diff
@@ -0,0 +1,21 @@
+import io.micronaut.jms.annotations.JMSListener
+import io.micronaut.jms.sqs.configuration.SqsConfiguration
+import io.micronaut.messaging.annotation.MessageBody
+import com.example.DemoConsumer
```
remove this import

burtbeckwith commented on guides/micronaut-jms-aws-sqs/kotlin/src/main/kotlin/example/micronaut/DemoConsumer.kt:
```diff
@@ -0,0 +1,21 @@
+class DemoConsumer {
+
+    @Queue(value = "demo_queue", concurrency = "1-3")
+    fun receive(@MessageBody body: String?) {
```
remove `?` unless null body is allowed

burtbeckwith commented on guides/micronaut-jms-aws-sqs/java/src/main/java/example/micronaut/DemoConsumer.java:
```diff
@@ -0,0 +1,19 @@
+
+@JMSListener(CONNECTION_FACTORY_BEAN_NAME)
+public class DemoConsumer {
+    private static Logger LOGGER = LoggerFactory.getLogger(DemoConsumer.class);
```
should be `final`

burtbeckwith commented on guides/micronaut-jms-aws-sqs/groovy/src/main/groovy/example/micronaut/DemoConsumer.groovy:
```diff
@@ -0,0 +1,19 @@
+
+import static io.micronaut.jms.sqs.configuration.SqsConfiguration.CONNECTION_FACTORY_BEAN_NAME
+
+@JMSListener(CONNECTION_FACTORY_BEAN_NAME)
```
If possible all Groovy classes should be annotated with `@CompileStatic`. I've had problems using this with controllers, but generally it works and will generate more efficient bytecode

burtbeckwith commented on guides/micronaut-jms-aws-sqs/kotlin/src/main/kotlin/example/micronaut/DemoProducer.kt:
```diff
@@ -0,0 +1,13 @@
+import io.micronaut.jms.sqs.configuration.SqsConfiguration
+import io.micronaut.messaging.annotation.MessageBody
+
+@JMSProducer(SqsConfiguration.CONNECTION_FACTORY_BEAN_NAME)
```
use static import to be consistent

burtbeckwith commented on guides/micronaut-jms-aws-sqs/src/main/resources/logback.xml:
```diff
@@ -0,0 +1,15 @@
+<configuration>
```
delete this file - it's the same as the file that's generated by the `create-app` step during the build

burtbeckwith commented on guides/micronaut-jms-aws-sqs/micronaut-jms-aws-sqs.adoc:
```diff
@@ -0,0 +1,121 @@
+
+== Micronaut JMS SQS Dependency
+
+To use the Micronaut JMS SQS integration, add the following dependency:
```
The dependency is added by the `jms-sqs` feature in the create-app command. Either remove this section, or change it to point out to the user that the feature adds this dependency. That's probably not needed though - they can see the effect of using the feature by clicking the diff link in the "If you have an existing Micronaut application ..." section

sbodvanski replied:
Good point

burtbeckwith reviewed (changes requested):
looks great, only minor changes needed

burtbeckwith commented on guides/micronaut-jms-aws-sqs/metadata.json:
```diff
@@ -0,0 +1,14 @@
+  "apps": [
+    {
+      "name": "default",
+      "features": ["jms-sqs"]
```
add `graalvm` feature

sbodvanski committed:
Guide: Connect a Micronaut JMS Application to an AWS SQS Queue

alvarosanchez reviewed (changes requested):
This guide is using AWS SDK v1's `com.amazonaws.services.sqs.AmazonSQS`. It should also (or only) demonstrate how to use v2's `software.amazon.awssdk.services.sqs.SqsClient`, which is [preconfigured](https://micronaut-projects.github.io/micronaut-aws/latest/guide/#sqs) by our `micronaut-aws-sdk-v2`

sbodvanski committed:
Guide: Connect a Micronaut JMS Application to an AWS SQS Queue

sbodvanski commented:
> AWS SDK v1's

I think that the current SQS sub-module in micronaut-jms supports only AWS SDK v1's (https://github.com/micronaut-projects/micronaut-jms/blob/master/jms-sqs/build.gradle). My understanding is that to support SDK V2, we will have to add a new module to micronaut-jms. Probably, that should be addressed in a separate ticket.

@alvarosanchez please take a look and let me know if I overlooked something, and if there is a way to work around the version restriction in the micronaut-jms module.

sbodvanski committed:
Guide: Connect a Micronaut JMS Application to an AWS SQS Queue

alvarosanchez commented:
This is so funny, because a random guy already noticed that some time ago: https://github.com/micronaut-projects/micronaut-jms/issues/35

😅

sbodvanski commented:
> micronaut-projects/micronaut-jms#35
:-D

sbodvanski committed:
Guide: Connect a Micronaut JMS Application to an AWS SQS Queue

Max supported Java version by AWS SDK 1.x is 16 (https://github.com/aws/aws-sdk-java#maintenance-and-support-for-java-versions)

sbodvanski commented:
> This is so funny, because a random guy already noticed that some time ago: [micronaut-projects/micronaut-jms#35](https://github.com/micronaut-projects/micronaut-jms/issues/35)
> 
> 😅

Do you guys think it is still worth finishing this guide (JMS AWS SDK v1.x integration)? When support for SDK v2 is added to micronaut-jms, we could provide the V2 guide.

alvarosanchez commented on guides/micronaut-jms-aws-sqs/java/src/main/java/example/micronaut/SqsClientFactory.java:
```diff
@@ -0,0 +1,24 @@
+import jakarta.inject.Singleton;
+
+@Factory  // <1>
+@Requires(notEnv = Environment.TEST)
```
This will confuse readers, but actually you don't need it. Simply remove `@Requires` from here. Will add an inline comment in the test factory.

alvarosanchez commented on guides/micronaut-jms-aws-sqs/groovy/src/test/groovy/example/micronaut/MicronautguideSpec.groovy:
```diff
@@ -0,0 +1,9 @@
+
+
+class MicronautguideSpec extends Specification {
+    // Localstack based tests are not supported.
```
If Spock tests are not to be implemented, simply disable them in metadata.json, eg:

https://github.com/micronaut-projects/micronaut-guides/blob/master/guides/micronaut-aws-lambda-eventbridge-event/metadata.json#L9

alvarosanchez commented on guides/micronaut-jms-aws-sqs/java/src/test/java/example/micronaut/TestSqsClientFactory.java:
```diff
@@ -0,0 +1,30 @@
+import java.util.Optional;
+
+@Factory
+@Requires(env = Environment.TEST)
```
This is redundant since the class belong to the test sources, so no way it can be loaded outside tests.

Instead, add `@Replaces(SqsClientFactory.class)`

alvarosanchez commented:
> Do you guys think it is still worth finishing this guide (JMS AWS SDK v1.x integration)? When support for SDK v2 is added to micronaut-jms, we could provide the V2 guide.

Sure! Just address pending feedback, and then we can assign it to Sergio for publishing

sbodvanski committed:
Guide: Connect a Micronaut JMS Application to an AWS SQS Queue

sbodvanski commented:
When I download the generated sources from the guide, I can compile them in IDE and run the app as well as the tests. However, the micronaut-guides CI build is failing because of:
```
Executing 'micronaut-jms-aws-sqs-maven-java' tests
Error:  COMPILATION ERROR : 
Error:  /home/runner/work/micronaut-guides/micronaut-guides/build/code/micronaut-jms-aws-sqs/micronaut-jms-aws-sqs-maven-java/src/main/java/example/micronaut/SqsClientFactory.java:[3,29] package com.amazonaws.regions does not exist
Error:  Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.10.1:compile (default-compile) on project micronautguide: Compilation failure
Error:  /home/runner/work/micronaut-guides/micronaut-guides/build/code/micronaut-jms-aws-sqs/micronaut-jms-aws-sqs-maven-java/src/main/java/example/micronaut/SqsClientFactory.java:[3,29] package com.amazonaws.regions does not exist
```

Not sure what I have missed here. The missing package referred to in the compilation error above is from the `com.amazonaws:aws-java-sdk-core` library.  I've found that it has been already added here: https://github.com/micronaut-projects/micronaut-guides/blob/master/buildSrc/src/main/resources/pom.xml#L21.

Do I need to introduce this lib or reintroduce as part of the new feature in the 'buildSrc/src/main/java/io/micronaut/guides/feature' ?

alvarosanchez commented:
That pom.xml is only for having automatically upgradeable dependency versions, but it alone won't make it part of the generated build

sbodvanski committed:
Guide: Connect a Micronaut JMS Application to an AWS SQS Queue

sbodvanski commented:
Have an additional question regarding the micronaut-guides  test sources compile task. Not sure why compile fails for Kotlin with the following error:

```
Error:  Failed to execute goal org.jetbrains.kotlin:kotlin-maven-plugin:1.6.10:test-compile (test-compile) on project micronautguide: Compilation failure: Compilation failure: 

Error:  /home/runner/work/micronaut-guides/micronaut-guides/build/code/micronaut-jms-aws-sqs/micronaut-jms-aws-sqs-maven-kotlin/src/test/kotlin/example/micronaut/MicronautguideTest.kt:[38,16] Unresolved reference: LocalStack
Error:  /home/runner/work/micronaut-guides/micronaut-guides/build/code/micronaut-jms-aws-sqs/micronaut-jms-aws-sqs-maven-kotlin/src/test/kotlin/example/micronaut/MicronautguideTest.kt:[43,9] Unresolved reference: LocalStack
```

What did I overlook here?

The same approach having `groovy` test source folder as well as `java` test source folder, which contains the LocalStack common helper class works for the Groovy task.

sbodvanski committed:
Guide: Connect a Micronaut JMS Application to an AWS SQS Queue

alvarosanchez commented on buildSrc/src/main/java/io/micronaut/guides/feature/Localstack.java:
```diff
@@ -26,6 +26,6 @@ public void processSelectedFeatures(FeatureContext featureContext) {
     public void apply(GeneratorContext generatorContext) {
         super.apply(generatorContext);
-        addDependency(generatorContext, "aws-java-sdk-core", Scope.TEST);
+        addDependency(generatorContext, "aws-java-sdk-core");
```
Is this change intentional?

sbodvanski replied:
Yes, otherwise the ci build was failing with the compilation error:

```
Executing 'micronaut-jms-aws-sqs-maven-java' tests
Error:  COMPILATION ERROR : 
Error:  /home/runner/work/micronaut-guides/micronaut-guides/build/code/micronaut-jms-aws-sqs/micronaut-jms-aws-sqs-maven-java/src/main/java/example/micronaut/SqsClientFactory.java:[3,29] package com.amazonaws.regions does not exist
Error:  Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.10.1:compile (default-compile) on project micronautguide: Compilation failure
Error:  /home/runner/work/micronaut-guides/micronaut-guides/build/code/micronaut-jms-aws-sqs/micronaut-jms-aws-sqs-maven-java/src/main/java/example/micronaut/SqsClientFactory.java:[3,29] package com.amazonaws.regions does not exist 
```

sdelamo replied:
This should not be necessary. This will add local stack to the main classpath.

sbodvanski replied:
Actually, It does not add the localstack dependency to the main classpath, it just changes the scope of the `aws-java-sdk-core `dependency. There are no issues when the demo source is generated and imported to IDE. 
However, when I revert the change, the problem occurs during the micronut-guide project build. Tests cannot be compiled because `aws-java-sdk-core` artifacts are missing.

@sdelamo what do you suggest?

alvarosanchez commented on guides/micronaut-jms-aws-sqs/micronaut-jms-aws-sqs.adoc:
```diff
@@ -0,0 +1,133 @@
+* An https://aws.amazon.com/[AWS account] with:
+** An IAM user with enough permissions to create and manage a queue instances in SQS.
+** The AWS CLI configured to use the IAM user above.
+
```
```suggestion

common:aws.adoc[]

```

alvarosanchez commented:
@sbodvanski please click on "Resolve conversation" for the feedback items that you already addressed, otherwise it's impossible for reviewers to tell what is done and what is still pending

sbodvanski commented:
> @sbodvanski please click on "Resolve conversation" for the feedback items that you already addressed, otherwise it's impossible for reviewers to tell what is done and what is still pending

Will do. Thank you for pointing it out.

sbodvanski committed:
Guide: Connect a Micronaut JMS Application to an AWS SQS Queue

sdelamo commented on guides/micronaut-jms-aws-sqs/micronaut-jms-aws-sqs.adoc:
```diff
@@ -0,0 +1,133 @@
+common:create-app-features.adoc[]
+
+== Create an application
+Let’s create a set of components that will use the Micronaut JMS to integrate with Amazon SQS. This will allow us to send and receive messages from the SQS. The SQS is a reliable, highly-scalable hosted queue for storing messages as they travel between applications or microservices. Amazon SQS moves data between distributed application components and helps you decouple these components.
```
Missing link to SQS and use a blockquote for content which is quoting SQS documentation.

sbodvanski replied:
Added

sdelamo commented on guides/micronaut-jms-aws-sqs/metadata.json:
```diff
@@ -0,0 +1,17 @@
+  "authors": ["Slavko Bodvanski"],
+  "tags": ["jms", "messaging", "aws"],
+  "categories": ["Messaging"],
+  "maximumJavaVersion": 16,
```
why fix maximumJavaVersion?

sbodvanski replied:
Until we add a new module to micronaut-jms to support Java AWS SDK v2's, it seems that we are limited to Java version below 17:
https://github.com/aws/aws-sdk-java#maintenance-and-support-for-java-versions

sdelamo commented on guides/micronaut-jms-aws-sqs/java/src/main/java/example/micronaut/DemoController.java:
```diff
@@ -0,0 +1,20 @@
+    }
+
+    @Get("/demo")  // <3>
+    public void publishDemoMessages() {
```
The controller which has a side effect it publishes to a Queue. It should be mapped to POST instead of Get. 

I would annotate the method with `@Status(NO_CONENT)` and respond a 204 instead 200.

sbodvanski replied:
Make sense. will change that.

sdelamo commented on guides/micronaut-jms-aws-sqs/groovy/src/main/groovy/example/micronaut/DemoController.groovy:
```diff
@@ -0,0 +1,21 @@
+        this.demoProducer = demoProducer
+    }
+
+    @Get("/demo")  // <3>
```
The controller which has a side effect it publishes to a Queue. It should be mapped to POST instead of Get. 

I would annotate the method with `@Status(NO_CONENT)` and respond a 204 instead 200.

sdelamo commented on guides/micronaut-jms-aws-sqs/kotlin/src/main/kotlin/example/micronaut/DemoController.kt:
```diff
@@ -0,0 +1,13 @@
+@Controller  // <1>
+class DemoController(private val demoProducer: DemoProducer) {  // <2>
+
+    @Get("/demo")  // <3>
```
The controller which has a side effect it publishes to a Queue. It should be mapped to POST instead of Get. 

I would annotate the method with `@Status(NO_CONENT)` and respond a 204 instead 200.

sdelamo commented on guides/micronaut-jms-aws-sqs/groovy/src/test/groovy/example/micronaut/MicronautguideTest.groovy:
```diff
@@ -0,0 +1,46 @@
+@Testcontainers(disabledWithoutDocker = true)
+@TestInstance(TestInstance.Lifecycle.PER_CLASS)
+@MicronautTest
+class MicronautguideTest implements TestPropertyProvider {
```
why not write the groovy guide tests with Spock instead of Junit5 as we do for every guide?

sbodvanski replied:
Have replaced it with the Spock test implementation.

sdelamo commented on guides/micronaut-jms-aws-sqs/metadata.json:
```diff
@@ -0,0 +1,17 @@
+  "tags": ["jms", "messaging", "aws"],
+  "categories": ["Messaging"],
+  "maximumJavaVersion": 16,
+  "testFramework": "junit",
```
it would be good to remove this and write Groovy tests with Spock.

sdelamo commented on guides/micronaut-jms-aws-sqs/groovy/src/test/groovy/example/micronaut/MicronautguideTest.groovy:
```diff
@@ -0,0 +1,46 @@
+    void testItWorks() {
+        int messageCount = demoConsumer.getMessageCount()
+        assert messageCount == 0
+        demoController.publishDemoMessages()
```
I would prefer if we injected a HTTP Client and invoke the controller from the outside. 

Replace 

```
@Inject
DemoController demoController
```

with 

```
@Inject
@Client("/")
HttpClient httpClient
````

replace: 

`demoController.publishDemoMessages()`

with: 

`httpClient.toBlocking().exchange(HttpRequest.POST("/demo", Collections.emptyMap())`

sbodvanski replied:
Will do it.

sdelamo commented on guides/micronaut-jms-aws-sqs/java/src/test/java/example/micronaut/MicronautguideTest.java:
```diff
@@ -0,0 +1,49 @@
+class MicronautguideTest implements TestPropertyProvider {
+
+    @Inject
+    DemoController demoController;
```
I would prefer if we injected a HTTP Client and invoke the controller from the outside. 

Replace 

```
@Inject
DemoController demoController
```

with 

```
@Inject
@Client("/")
HttpClient httpClient
````

replace: 

`demoController.publishDemoMessages()`

with: 

`httpClient.toBlocking().exchange(HttpRequest.POST("/demo", Collections.emptyMap())`

sdelamo commented on guides/micronaut-jms-aws-sqs/kotlin/src/test/kotlin/example/micronaut/MicronautguideTest.kt:
```diff
@@ -0,0 +1,46 @@
+internal class MicronautguideTest : TestPropertyProvider {
+
+    @Inject
+    lateinit var demoController: DemoController
```
I would prefer if we injected a HTTP Client and invoke the controller from the outside. 

Replace 

```
@Inject
DemoController demoController
```

with 

```
@Inject
@Client("/")
HttpClient httpClient
````

replace: 

`demoController.publishDemoMessages()`

with: 

`httpClient.toBlocking().exchange(HttpRequest.POST("/demo", Collections.emptyMap())`

sdelamo commented on guides/micronaut-jms-aws-sqs/micronaut-jms-aws-sqs.adoc:
```diff
@@ -0,0 +1,133 @@
+common:create-app-features.adoc[]
+
+== Create an application
+Let’s create a set of components that will use the Micronaut JMS to integrate with Amazon SQS. This will allow us to send and receive messages from the SQS. The SQS is a reliable, highly-scalable hosted queue for storing messages as they travel between applications or microservices. Amazon SQS moves data between distributed application components and helps you decouple these components.
```
```suggestion
Let’s create a set of components that will use the https://micronaut-projects.github.io/micronaut-jms/latest/guide/[Micronaut JMS] to send and receive messages from https://docs.aws.amazon.com/it_it/AWSSimpleQueueService/latest/APIReference/Welcome.html[AWS SQS]

____
Amazon SQS is a reliable, highly-scalable hosted queue for storing messages as they travel between applications or microservices. Amazon SQS moves data between distributed application components and helps you decouple these components.
____
```

sbodvanski committed:
Guide: Connect a Micronaut JMS Application to an AWS SQS Queue
